## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [933]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [934]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [935]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [936]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(f"The different data types are: {df.dtypes}")

df.columns

df.count






The different data types are: first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


<bound method DataFrame.count of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id 

In [937]:
# What three item categories had the most entries?
category_count = df['category'].value_counts()
print(category_count.head(3))
df.dtypes



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [938]:
# For the category with the most entries,
# which subcategory had the most entries?
consumables_df = df[df["category"]== 'consumables']
subcategory_count = consumables_df['subcategory'].value_counts()
top_subcategory = subcategory_count.max()
print(subcategory_count.head(1))

subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [939]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [940]:
# Store the client ids of those top 5 clients in a list.
top_clients_ids = top_clients.index.tolist()



In [941]:
# How many total units (the qty column) did the
# client with the most entries order order?
total_units_top = df[df['client_id']==top_clients_ids[0]]['qty'].sum()
total_units_top


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [942]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [943]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def shipping_caculate(weight,qty):
    if weight > 50:
        return qty*weight*7
    else:
        return qty*weight*10

df['shipping_price'] = df.apply(lambda x: shipping_caculate(x['unit_weight'],x['qty']),axis=1)


In [944]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df['total_price'] 



0        134420.107500
1           914.258625
2          1291.859400
3          1543.571400
4         14407.890000
             ...      
54634      3808.225575
54635     16615.537525
54636     55661.236250
54637     22254.836800
54638     26706.173425
Name: total_price, Length: 54639, dtype: float64

In [945]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']


In [946]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [947]:
# Check your work using the totals above

order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    actual_total = df[df['order_id'] == order_id]['total_price'].sum()
    print(f"Order {order_id} total: ${actual_total:.2f}")



Order 2742071 total: $158105.65
Order 2173913 total: $180488.06
Order 6128929 total: $996593.76


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [948]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients_data = df[df['client_id'].isin(top_clients_ids)]
client_aggregates = top_clients_data.groupby('client_id').agg({
    'qty': 'sum', 
    'shipping_price': 'sum',  
    'total_price': 'sum',  
    'profit': 'sum'  
}).reset_index()

# Sort by total profit
client_aggregates.sort_values(by='profit', ascending=False, inplace=True)
client_aggregates

client_id     qty  shipping_price   total_price        profit
0      24741  239862       7318757.5  8.466399e+07  3.678278e+07
2      38378   73667       4895361.6  1.450805e+07  3.407426e+06
4      66037   43018       1989501.5  1.090884e+07  3.310009e+06
3      46820   75768       2284215.0  1.048972e+07  2.799759e+06
1      33615   64313       2608560.3  9.228995e+06  2.274105e+06

In [949]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
client_aggregates.rename(columns={
    'qty': 'total_units_purchased',
    'shipping_price': 'total_shipping_price',
    'total_price': 'total_revenue',
    'profit': 'total_profit'
}, inplace=True)


client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862             7318757.5   8.466399e+07   
2      38378                  73667             4895361.6   1.450805e+07   
4      66037                  43018             1989501.5   1.090884e+07   
3      46820                  75768             2284215.0   1.048972e+07   
1      33615                  64313             2608560.3   9.228995e+06   

   total_profit  
0  3.678278e+07  
2  3.407426e+06  
4  3.310009e+06  
3  2.799759e+06  
1  2.274105e+06

In [950]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

client_aggregates['total_shipping_price'] = client_aggregates['total_shipping_price'] / 1_000_000 
client_aggregates['total_revenue'] = client_aggregates['total_revenue'] / 1_000_000  
client_aggregates['total_profit'] = client_aggregates['total_profit'] / 1_000_000 
client_aggregates['total_shipping_price'] = client_aggregates['total_shipping_price'].map('${:,.2f}M'.format)
client_aggregates['total_revenue'] = client_aggregates['total_revenue'].map('${:,.2f}M'.format)
client_aggregates['total_profit'] = client_aggregates['total_profit'].map('${:,.2f}M'.format)
client_aggregates.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Total Units',
    'total_shipping_price': 'Shipping Costs (M$)',
    'total_revenue': 'Total Revenue (M$)',
    'total_profit': 'Total Profit (M$)'
}, inplace=True)
client_aggregates

Client ID  Total Units Shipping Costs (M$) Total Revenue (M$)  \
0      24741       239862              $7.32M            $84.66M   
2      38378        73667              $4.90M            $14.51M   
4      66037        43018              $1.99M            $10.91M   
3      46820        75768              $2.28M            $10.49M   
1      33615        64313              $2.61M             $9.23M   

  Total Profit (M$)  
0           $36.78M  
2            $3.41M  
4            $3.31M  
3            $2.80M  
1            $2.27M

In [951]:
# Sort the updated data by "Total Profit" form highest to lowest
